compute_tools/src/extension_server.rs

```diff
@@ -0,0 +1,242 @@
+    pgbin: &str,
+    pg_version: &str,
+    enabled_extensions: &[String],
+    build_tag: &str,
+) -> Result<HashMap<String, RemotePath>> {
+    let local_sharedir = Path::new(&get_pg_config("--sharedir", pgbin)).join("extension");
+    let index_path = format!("{build_tag}/{pg_version}/ext_index.json");
+    let index_path = RemotePath::new(Path::new(&index_path)).context("error forming path")?;
+    info!("download ext_index.json from: {:?}", &index_path);
+
+    let mut download = better_download(remote_storage, &index_path).await?;
+    let mut ext_idx_buffer = Vec::new();
+    download
+        .download_stream
+        .read_to_end(&mut ext_idx_buffer)
+        .await?;
+    let ext_index_str = str::from_utf8(&ext_idx_buffer).expect("error parsing json");
+    let ext_index_full: Value = serde_json::from_str(ext_index_str)?;
+    let ext_index_full = ext_index_full.as_object().context("error parsing json")?;
+    info!("ext_index: {:?}", &ext_index_full);
+
+    info!("enabled_extensions: {:?}", enabled_extensions);
+    let mut ext_remote_paths = HashMap::new();
+    let mut file_create_tasks = Vec::new();
+    for extension in enabled_extensions {
+        let ext_data = ext_index_full[extension]
+            .as_object()
+            .context("error parsing json")?;
+
+        let control_files = ext_data["control_data"]
+            .as_object()
+            .context("error parsing json")?;
+        for (control_file, control_contents) in control_files {
+            let control_path = local_sharedir.join(control_file);
+            let control_contents_str = control_contents.as_str().context("error parsing json")?;
+            info!("writing file {:?}{:?}", control_path, control_contents);
+            file_create_tasks.push(tokio::fs::write(control_path, control_contents_str));
+        }
+
+        let ext_archive_path = ext_data["archive_path"]
+            .as_str()
+            .context("error parsing json")?;
+        ext_remote_paths.insert(
+            extension.to_string(),
+            RemotePath::from_string(ext_archive_path)?,
+        );
+    }
+    let results = join_all(file_create_tasks).await;
+    for result in results {
+        result?;
+    }
+    Ok(ext_remote_paths)
+}
+
+// download the archive for a given extension,
+// unzip it, and place files in the appropriate locations (share/lib)
+pub async fn download_extension(
+    ext_name: &str,
+    ext_path: &RemotePath,
+    remote_storage: &GenericRemoteStorage,
+    pgbin: &str,
+) -> Result<()> {
+    info!("Download extension {:?} from {:?}", ext_name, ext_path);
+    let mut download = better_download(remote_storage, ext_path).await?;
+    let mut download_buffer = Vec::new();
+    download
+        .download_stream
+        .read_to_end(&mut download_buffer)
+        .await?;
+    let mut decoder = Decoder::new(download_buffer.as_slice())?;
+    let mut decompress_buffer = Vec::new();
+    decoder.read_to_end(&mut decompress_buffer)?;
```

Comment:
How large are these files? Looks like we're loading them into memory

Reply:
I think postgis, one of the larger extensions, is ~10MB

Reply:
Ok so probably won't OOM, unless there's a large number of extensions (idk how many people actually use). Maybe add a `TODO decompressing into memory can be avoided` so if there's a problem it's obvious where to look. Might be faster to avoid allocations too, but hard to tell

Reply:
I'll add `TODO decompressing into memory can be avoided`